src/hb-utils.c

```diff
@@ -0,0 +1,138 @@
+  const char * filename;
+  int face_index = 0;
+  hb_blob_t* blob;
+  hb_face_t* face;
+  hb_font_t* font;
+  unsigned int upem;
+
+  luaL_checktype(L, index, LUA_TTABLE);
+
+  lua_getfield(L, index, "hbFont");
+  if (lua_islightuserdata(L, -1)) { return lua_touserdata(L, -1); }
+
+  lua_getfield(L, index, "filename");
+  if (lua_isstring(L, -1)) { filename = lua_tostring(L, -1); }
+  else { luaL_error(L, "No font filename supplied to get_glyph_dimensions"); }
+
+  lua_getfield(L, index, "index");
+  if (lua_isnumber(L, -1)) { face_index = lua_tointeger(L, -1); }
+
+  blob = hb_blob_create_from_file(filename);
+  face = hb_face_create(blob, face_index);
+  font = hb_font_create(face);
+  upem = hb_face_get_upem(face);
+  hb_font_set_scale(font, upem, upem);
+
+  hb_ot_font_set_funcs(font);
+
+#if HB_VERSION_ATLEAST(3, 3, 0)
+  if (hb_ot_var_has_data(face)) {
+    hb_ot_var_axis_info_t* axes;
+    unsigned int nAxes;
+    unsigned int nCoords;
+    const float* coords;
+    float* newCoords;
+
+    /* Get font axes */
+    nAxes = hb_ot_var_get_axis_infos(face, 0, NULL, NULL);
+    axes = malloc(nAxes * sizeof(hb_ot_var_axis_info_t));
+    hb_ot_var_get_axis_infos(face, 0, &nAxes, axes);
+
+    /* Get existing variation coords, e.g. from named instance */
+    coords = hb_font_get_var_coords_design(font, &nCoords);
+
+    /* Set up new variation coords */
+
+    /* First copy existing coords (e.g. from named instance) or use axis
+     * default value. */
+    newCoords = malloc(nAxes * sizeof(float));
+    for (unsigned i = 0; i < nAxes; i++) {
+      if (i < nCoords)
+        newCoords[i] = coords[i];
+      else
+        newCoords[i] = axes[i].default_value;
+    }
+
+    /* Then set ‘opsz’ axis to point size if present, explicitly set value will
+     * override this below */
+    lua_getfield(L, index, "pointsize");
+    if (lua_isnumber(L, -1)) {
+      double point_size = luaL_checknumber(L, -1);
+      for (unsigned i = 0; i < nAxes; i++) {
+        if (axes[i].tag == HB_TAG('o', 'p', 's', 'z'))
+          newCoords[i] = point_size;
+      }
+    }
+
+    /* Finally use any explicitly set variations */
+    lua_getfield(L, index, "variations");
+    if (lua_isstring(L, -1)) {
+      const char* variationstring = lua_tostring(L, -1);
+      unsigned int nVariations = 0;
+      hb_variation_t* variations = scan_variation_string(variationstring, &nVariations);
+      if (variations) {
+        for (unsigned i = 0; i < nVariations; i++) {
+          for (unsigned j = 0; j < nAxes; j++) {
+            if (variations[i].tag == axes[j].tag)
+              newCoords[i] = variations[i].value;
+          }
+        }
+        free(variations);
+      }
+    }
+
+    hb_font_set_var_coords_design(font, newCoords, nAxes);
+
+    free(axes);
+    free(newCoords);
+  }
+#endif
+
+  hb_face_destroy(face);
+  hb_blob_destroy(blob);
+
+  lua_pushlightuserdata(L, font);
+  lua_setfield(L, index, "hbFont");
```

Comment:
We never call `hb_font_destroy()`, but since the fonts are cached and never garbage collected, this seems to not be an issue, this will be released when the program exits anyway. It should be easy to fix though (adding a method to be called from `SILE.font.finish()`.